Library/Homebrew/extend/os/linux/diagnostic.rb

```diff
@@ -143,6 +143,12 @@ def check_gcc_dependent_linkage
           return if gcc_dependents.empty?
 
           badly_linked = gcc_dependents.select do |dependent|
+            # Keg.new() may raise an error if it is not a directory.
+            # As the result `brew doctor` may display `Error: <keg> is not a directory`
+            # instead of proper `doctor` information.
+            # There are other checks that test that, we can skip broken kegs.
+            next if !dependent.prefix.exist? || !dependent.prefix.directory?
+
             keg = Keg.new(dependent.prefix)
```

Comment:
Let's do this so we don't have to assume `dependent` is `optlinked`:
```suggestion
            dependent_prefix = dependent.any_installed_prefix
            # Keg.new() may raise an error if it is not a directory.
            # As the result `brew doctor` may display `Error: <keg> is not a directory`
            # instead of proper `doctor` information.
            # There are other checks that test that, we can skip broken kegs.
            next if !dependent_prefix.exist? || !dependent_prefix.directory?

            keg = Keg.new(dependent_prefix)
```